// likec4/dist/_chunks/plugin.mjs
import{t as createBirpc}from"./libs/birpc.mjs";import{E as e,F as t$5,K as n,O as e$1,S as e$2,U as n$2,X as t$1,d as t$4,o as t$2,r as n$1,v as t,y as t$3}from"./libs/remeda.mjs";import{a as GraphvizWasmAdapter,f as handleInitOptions,i as QueueGraphvizLayoter,l as NoFileSystem,m as logger$1,n as createLanguageServices$1,o as WithFileSystem,s as WithLikeC4ManualLayouts,t as GraphvizBinaryAdapter,u as NoLikeC4ManualLayouts,y as loggable}from"./GraphvizBinaryAdapter.mjs";import{t as defu}from"./libs/defu.mjs";import{r as resolve,t as basename}from"./libs/pathe.mjs";import{t as u}from"./libs/tinyrainbow.mjs";import{t as lib}from"./libs/json5.mjs";import{a as expandToNode,i as toString,n as CompositeGeneratorNode,r as NL,t as joinToNode}from"./libs/langium.mjs";import{o as joinURL}from"./libs/ufo.mjs";import{LikeC4Model}from"@likec4/core/model";import{invariant,isNonEmptyArray,memoizeProp,nonexhaustive}from"@likec4/core";import{pathToFileURL}from"url";import{RichText as RichText$1}from"@likec4/core/types";import{compareNatural}from"@likec4/core/utils";function createLanguageServices(r){let i=logger$1.getChild(`lang`),a=defu(r,{useFileSystem:!0,manualLayouts:!0,watch:!1,graphviz:`wasm`}),o=a.graphviz===`binary`;return i.info(`${u.dim(`layout`)} ${o?`binary`:`wasm`}`),createLanguageServices$1({...a.useFileSystem?{...WithFileSystem(a.watch),...a.manualLayouts?WithLikeC4ManualLayouts:NoLikeC4ManualLayouts}:{...NoFileSystem,...NoLikeC4ManualLayouts}},{likec4:{Layouter:()=>new QueueGraphvizLayoter({graphviz:o?new GraphvizBinaryAdapter:new GraphvizWasmAdapter})}})}async function fromWorkspace(r,i){let a=resolve(r);return memoizeProp(globalThis,`likec4:`+a,async()=>{let r=logger$1.getChild(`lang`),o=defu(i,{useFileSystem:!0,manualLayouts:!0,watch:!1});if(o.mcp)throw Error(`MCP server is not supported in this build`);let s=createLanguageServices(o),c={name:basename(a),uri:pathToFileURL(a).toString()},l=s.shared.workspace.WorkspaceManager;r.info(`${u.dim(`workspace:`)} ${a}`),l.initialize({capabilities:{},processId:null,rootUri:null,workspaceFolders:[c]}),await l.initializeWorkspace([c]);let d=s.shared.workspace.LangiumDocuments.userDocuments.toArray();if(d.length===0)throw r.error(`no LikeC4 sources found`),Error(`no LikeC4 sources found`);return r.info(`${u.dim(`workspace:`)} found ${d.length} source files`),handleInitOptions(s,logger$1,i)})}const logger=logger$1.getChild(`vite`);async function calcAdhocView({logger:r,likec4:i},a){r.info([u.green(`adhoc:view`),u.dim(`project:`),a.projectId].join(` `));let o=await i.views.adhocView(a.predicates,a.projectId);return r.info([u.green(`adhoc:view`),`✅`].join(` `)),o}function splitErrorMessage(r){let i=r.split(`
`),a=/^\s+at\s+/,o=i.findIndex(r=>a.test(r));return o===-1&&(o=1),{message:i.slice(0,o).join(`
`),stack:i.slice(o).join(`
`)}}function sendError(r,{name:i,error:a}){r.hot.send({type:`error`,err:{...splitErrorMessage(a),name:i??`LikeC4PluginError`,plugin:`vite-plugin-likec4`}})}async function updateView({logger:r,likec4:i},a){r.info([u.green(`view:onChange`),u.dim(`project`)+`:`,a.projectId,u.dim(`view`)+`:`,a.viewId,u.dim(`change`)+`:`,a.change.op].join(` `));let o=await i.editor.applyChange(a);if(!o.success){r.error(`Failed to apply view change:\n${o.error}`);let i=Error(o.error);throw i.stack=o.error,i}r.info([u.green(`view:onChange`),`✅`].join(` `))}function enablePluginRPC(i){let a=null,o=i.server;createBirpc({updateView:r=>updateView(i,r),calcAdhocView:r=>calcAdhocView(i,r)},{on:r=>o.hot.on(`likec4:rpc`,r),post:r=>o.hot.send(`likec4:rpc`,r),onFunctionError:(r,s)=>{i.logger.error(`RPC error in ${s}`,{error:r});let c=r.stack??r.message;try{a!==c&&(a=c,sendError(o,{name:r.name,error:c}))}catch(r){i.logger.error(`Failed to send error to client`,{error:r})}}})}const capitalizeFirstLetter$2=r=>r.charAt(0).toLocaleUpperCase()+r.slice(1),fqnName$2=r=>r.split(`.`).map(capitalizeFirstLetter$2).join(``),nodeName$2=r=>fqnName$2(r.parent?r.id.slice(r.parent.length+1):r.id),d2direction=({autoLayout:r})=>{switch(r.direction){case`TB`:return`down`;case`BT`:return`up`;case`LR`:return`right`;case`RL`:return`left`}},d2shape=({shape:r})=>{switch(r){case`queue`:case`cylinder`:case`rectangle`:case`document`:return r;case`person`:return`c4-person`;case`storage`:return`stored_data`;case`component`:case`bucket`:case`mobile`:case`browser`:return`rectangle`}};function generateD2(r){let a=r.$view,{nodes:o,edges:s}=a,c=new Map,l=(r,i)=>{let a=nodeName$2(r),s=(i?i+`.`:``)+a;c.set(r.id,s);let d=JSON.stringify(r.title),f=d2shape(r);return new CompositeGeneratorNode().append(a,`: {`,NL).indent({indentedChildren:i=>i.append(`label: `,d,NL).appendIf(f!==`rectangle`,`shape: `,f,NL).appendIf(r.children.length>0,NL,joinToNode(o.filter(i=>i.parent===r.id),r=>l(r,s))),indentation:2}).append(`}`,NL)},d=r=>new CompositeGeneratorNode().append(c.get(r.source),` -> `,c.get(r.target)).append(i=>r.label&&i.append(`: `,JSON.stringify(r.label)));return toString(new CompositeGeneratorNode().append(`direction: `,d2direction(a),NL,NL).append(joinToNode(o.filter(r=>e(r.parent)),r=>l(r),{appendNewLineIfNotEmpty:!0})).appendIf(s.length>0,NL,joinToNode(s,r=>d(r),{appendNewLineIfNotEmpty:!0})))}const capitalizeFirstLetter$1=r=>r.charAt(0).toLocaleUpperCase()+r.slice(1),fqnName$1=r=>r.split(`.`).map(capitalizeFirstLetter$1).join(``),nodeName$1=r=>fqnName$1(r.parent?r.id.slice(r.parent.length+1):r.id),toSingleQuotes=r=>r.replace(/\\?"/g,`'`),mmdshape=({shape:r,title:i})=>{let a=`label: ${JSON.stringify(i)}`;switch(r){case`queue`:return`@{ shape: horizontal-cylinder, ${a} }`;case`person`:return`@{ icon: "fa:user", shape: rounded, ${a} }`;case`storage`:return`@{ shape: disk, ${a} }`;case`cylinder`:return`@{ shape: cylinder, ${a} }`;case`mobile`:case`browser`:return`@{ shape: rounded, ${a} }`;case`bucket`:return`@{ shape: trap-t, ${a} }`;case`rectangle`:return`@{ shape: rectangle, ${a} }`;case`document`:return`@{ shape: doc, ${a} }`;case`component`:return`@{ shape: rectangle, ${a} }`;default:nonexhaustive(r)}};function generateMermaid(r){let a=r.$view,{nodes:o,edges:s}=a,c=new Map,l=(r,i)=>{let a=nodeName$1(r),s=(i?i+`.`:``)+a;c.set(r.id,s);let d=new CompositeGeneratorNode;if(r.children.length>0){let i=toSingleQuotes(r.title);d.append(`subgraph `,s,'["`',i,'`"]',NL).indent({indentedChildren:[joinToNode(o.filter(i=>i.parent===r.id),r=>l(r,s),{appendNewLineIfNotEmpty:!0})],indentation:2}).append(`end`,NL)}else d.append(s,mmdshape(r));return d},d=r=>new CompositeGeneratorNode().append(c.get(r.source),` -.`,r.label?' "`'+toSingleQuotes(r.label)+'`" .-':`-`,`> `,c.get(r.target));return toString(new CompositeGeneratorNode().append(`---`,NL,`title: ${JSON.stringify(toSingleQuotes(r.titleOrId))}`,NL,`---`,NL).append(`graph `,a.autoLayout.direction,NL).indent({indentedChildren:r=>{r.append(joinToNode(o.filter(r=>e(r.parent)),r=>l(r),{appendNewLineIfNotEmpty:!0})).appendIf(s.length>0,joinToNode(s,r=>d(r),{appendNewLineIfNotEmpty:!0}))},indentation:2}))}const capitalizeFirstLetter=r=>r.charAt(0).toLocaleUpperCase()+r.slice(1),fqnName=r=>r.split(/[.-]/).map(capitalizeFirstLetter).join(``),nodeName=r=>fqnName(r.parent?r.id.slice(r.parent.length+1):r.id),pumlColor=(r,i,a=`#3b82f6`)=>r?i(r)??a:a,pumlDirection=({autoLayout:r})=>{switch(r.direction){case`TB`:return`top to bottom`;case`BT`:return console.warn(`Bottom to top direction is not supported. Defaulting to top to bottom.`),`top to bottom`;case`LR`:return`left to right`;case`RL`:return console.warn(`Right to left direction is not supported. Defaulting to left to right.`),`left to right`}},pumlShape=({shape:r})=>{switch(r){case`queue`:case`rectangle`:case`person`:return r;case`storage`:case`cylinder`:return`database`;case`component`:return`component`;case`document`:case`mobile`:case`bucket`:case`browser`:return`rectangle`;default:nonexhaustive(r)}},escapeLabel=r=>e$1(r)?null:JSON.stringify(r).slice(1,-1).replace(/\\"/g,`"`);function generatePuml(r){let a=r.$view,o=r.$model.$styles.theme.colors,{nodes:c,edges:l}=a,d=r=>i=>i in o?o[i].elements[r]:void 0,f=r=>i=>i in o?o[i].relationships[r]:void 0,p=new Map,m=()=>new CompositeGeneratorNode().append(`title "`,r.titleOrId,`"`,NL).append(pumlDirection(a),` direction`,NL),h=()=>new CompositeGeneratorNode().append(`hide stereotype`,NL).append(`skinparam ranksep `,`60`,NL).append(`skinparam nodesep `,`30`,NL).append(`skinparam {`,NL).indent({indentedChildren:r=>r.append(`arrowFontSize `,`10`,NL).append(`defaultTextAlignment `,`center`,NL).append(`wrapWidth `,`200`,NL).append(`maxMessageSize `,`100`,NL).append(`shadowing `,`false`,NL),indentation:2}).append(`}`,NL),g=r=>{let i=pumlShape(r),a=fqnName(r.id);return new CompositeGeneratorNode().append(`skinparam `,i,`<<`,a,`>>`,`{`,NL).indent({indentedChildren:i=>i.append(`BackgroundColor `,pumlColor(r.color,d(`fill`)),NL).append(`FontColor `,pumlColor(r.color,d(`hiContrast`),`#FFFFFF`),NL).append(`BorderColor `,pumlColor(r.color,d(`stroke`)),NL),indentation:2}).append(`}`,NL)},_=r=>{let i=pumlShape(r),a=fqnName(r.id),o=escapeLabel(r.title)||nodeName(r),s=escapeLabel(r.technology);p.set(r.id,a);let c=RichText$1.from(r.description);return new CompositeGeneratorNode().append(i,` `).append(`"`).append(`==`,o).appendIf(!!s,`\\n<size:10>[`,s,`]</size>`).appendIf(c.nonEmpty,`\\n\\n`,escapeLabel(c.text)).append(`"`,` <<`,a,`>> `,`as `,a,NL)},v=r=>{let i=escapeLabel(r.title)||nodeName(r),a=fqnName(r.id);return p.set(r.id,a),new CompositeGeneratorNode().append(`rectangle "`,i,`" <<`,a,`>> as `,a,` {`,NL).indent({indentedChildren:i=>i.append(`skinparam `,`RectangleBorderColor<<`,a,`>> `,pumlColor(r.color,d(`fill`)),NL).append(`skinparam `,`RectangleFontColor<<`,a,`>> `,pumlColor(r.color,d(`fill`)),NL).append(`skinparam `,`RectangleBorderStyle<<`,a,`>> `,`dashed`,NL,NL).append(joinToNode(c.filter(i=>i.parent===r.id),r=>r.children.length>0?v(r):_(r))),indentation:2}).append(`}`,NL)},y=r=>{let i=r.technology||``,a=r.label||i,o=pumlColor(r.color,f(`line`),`#777777`),c=r=>`<color:${o}>${r.replaceAll(`"`,`'`)}`,l=new CompositeGeneratorNode().append(p.get(r.source),` .[`,o,`,thickness=2].> `,p.get(r.target));return(a||i)&&(l.append(` : `,a.split(`
`).map(r=>e$1(r)?r:c(r)).join(`\\n`)),i&&i!==a&&l.append(`\\n<size:8>[`,c(i),`]</size>`)),l.append(NL)};return toString(new CompositeGeneratorNode().append(`@startuml`,NL).append(m(),NL).append(h(),NL).append(joinToNode(c.filter(r=>r.children.length==0),r=>g(r),{appendNewLineIfNotEmpty:!0})).append(joinToNode(c.filter(r=>e(r.parent)),r=>r.children.length>0?v(r):_(r),{appendNewLineIfNotEmpty:!0})).appendIf(l.length>0,NL,joinToNode(l,r=>y(r),{appendNewLineIfNotEmpty:!0})).append(`@enduml`,NL))}function generateMatches(r,i=`.js`){return{matches:a=>{let{module:o,projectId:s}=a.match(/^likec4:plugin\/(?<projectId>.+)\/(?<module>.+)$/)?.groups??a.match(/^likec4:(?<module>.+)\/(?<projectId>.+)$/)?.groups??{};return!o||!s?null:(o.endsWith(i)&&(o=o.slice(0,-i.length)),o===r?s:null)},virtualId:a=>joinURL(`likec4:plugin`,a,r)+i}}const charMap={"<":`\\u003C`,">":`\\u003E`,"/":`\\u002F`,"\\":`\\\\`,"\b":`\\b`,"\f":`\\f`,"\n":`\\n`,"\r":`\\r`,"	":`\\t`,"\0":`\\0`,"\u2028":`\\u2028`,"\u2029":`\\u2029`};function escapeUnsafeChars(r){return r.replace(/[<>\b\f\n\r\t\0\u2028\u2029\\]/g,r=>charMap[r])}function generateCombinedProjects(r,i){return{id:`likec4:${r}`,virtualId:`likec4:plugin/${r}.js`,async load({logger:a,projects:o}){return a.info(u.dim(`generating likec4:${r}`)),`
export let ${i}Fn = {
${o.map(({id:i})=>{let a=escapeUnsafeChars(JSON.stringify(joinURL(`likec4:${r}`,i)));return`${JSON.stringify(i)}: () => import(${a})`}).join(`,
`)}
}      

export async function ${i}(projectId) {
  let fn = ${i}Fn[projectId]
  if (!fn) {
    const projects = Object.keys(${i}Fn)
    console.error('Unknown projectId: ' + projectId + ' (available: ' + projects + ')')
    if (projects.length === 0) {
      throw new Error('No projects found, invalid state')
    }
    projectId = projects[0]
    console.warn('Falling back to project: ' + projectId)
    fn = ${i}Fn[projectId]
  }
  return await fn()
}

if (import.meta.hot) {
  import.meta.hot.accept(md => {
    if (!import.meta.hot.data.$update) {
      import.meta.hot.data.$update = ${i}Fn
    }
    const update = md.${i}Fn
    if (update) {
      Object.assign(import.meta.hot.data.$update, update)
    } else {
      import.meta.hot.invalidate()
    }
  })
}
    `}}}function code$8(r){let i=new CompositeGeneratorNode;return i.appendTemplate`
    /******************************************************************************
     * This file was generated
     * DO NOT EDIT MANUALLY!
     ******************************************************************************/
    /* eslint-disable */

    export function d2Source(viewId) {
      switch (viewId) {
  `.appendNewLine().indent({indentation:4,indentedChildren(i){i.append(joinToNode([...r.views()],r=>expandToNode`
              case ${JSON.stringify(r.id)}: {
                return ${JSON.stringify(generateD2(r))}
              }
            `,{appendNewLineIfNotEmpty:!0})).appendTemplate`
        default: {
          throw new Error('Unknown viewId: ' + viewId)
        }
      `}}).append(NL,`  }`,NL).appendTemplate`
    }
    `.append(NL,NL),toString(i)}const projectD2Module={...generateMatches(`d2`),async load({likec4:r,project:i,logger:a}){return a.info(u.dim(`generating likec4:d2/${i.id}`)),code$8(await r.computedModel(i.id))}},d2Module=generateCombinedProjects(`d2`,`loadD2Sources`);function code$7(r){let i=new CompositeGeneratorNode;return i.appendTemplate`
    /******************************************************************************
     * This file was generated
     * DO NOT EDIT MANUALLY!
     ******************************************************************************/
    /* eslint-disable */

    export function dotSource(viewId) {
      switch (viewId) {
  `.appendNewLine().indent({indentation:4,indentedChildren(i){i.append(joinToNode(Object.keys(r),i=>expandToNode`
              case ${JSON.stringify(i)}: {
                return ${JSON.stringify(r[i].dot)}
              }
            `,{appendNewLineIfNotEmpty:!0})).appendTemplate`
          default: {
            throw new Error('Unknown viewId: ' + viewId)
          }
        `}}).append(NL,`  }`,NL).appendTemplate`
    }

    export function svgSource(viewId) {
      switch (viewId) {
    `.appendNewLine().indent({indentation:4,indentedChildren(i){i.append(joinToNode(Object.keys(r),i=>expandToNode`
              case ${JSON.stringify(i)}: {
                return ${JSON.stringify(r[i].svg)}
              }
            `,{appendNewLineIfNotEmpty:!0})).appendTemplate`
          default: {
            throw new Error('Unknown viewId: ' + viewId)
          }
        `}}).append(NL,`  }`,NL,`}`,NL,NL),toString(i)}const projectDotSourcesModule={...generateMatches(`dot`),async load({likec4:r,project:i,logger:a}){return a.info(u.dim(`generating likec4:dot/${i.id}`)),code$7(t(await r.views.viewsAsGraphvizOut(i.id),({id:r,svg:i,dot:a})=>[r,{dot:a,svg:i}]))}},dotModule=generateCombinedProjects(`dot`,`loadDotSources`);function code$6(r){let{imports:i,cases:a}=t$1(r.flatMap(r=>r.nodes.map(r=>r.icon)),n(r=>e$2(r)&&!(r.toLowerCase().startsWith(`http:`)||r.toLowerCase().startsWith(`https:`))),n$1(),t$2(compareNatural)).reduce((r,i,a)=>{let o=i.startsWith(`file:`),s=`Icon`+a.toString().padStart(2,`0`);if(o)return r.imports.push(`import ${s} from '${i}?inline'`),r.cases.push(`  '${i}': () => jsx('img', { src: ${s} })`),r;let[c,l]=i.split(`:`);return r.imports.push(`import ${s} from '@likec4/icons/${c}/${l}'`),r.cases.push(`  '${c}:${l}': ${s}`),r},{imports:[],cases:[]});return`
import { jsx } from 'react/jsx-runtime'
${i.join(`
`)}

const Icons = {
${a.join(`,
`)}
}
export function IconRenderer({ node, ...props }) {
  const IconComponent = Icons[node.icon ?? '']
  if (!IconComponent) {
    return null
  }
  return jsx(IconComponent, props)
}
`}const projectIconsModule={...generateMatches(`icons`,`.jsx`),async load({likec4:r,project:i,logger:a}){return a.info(u.dim(`generating likec4:icons/${i.id}`)),code$6(await r.views.computedViews(i.id))}},iconsModule={id:`likec4:icons`,virtualId:`likec4:plugin/icons.jsx`,async load({projects:r,logger:i}){return i.info(u.dim(`generating likec4:icons`)),`
import { jsx } from 'react/jsx-runtime'
import { lazy, Suspense } from 'react' 
export let ProjectIconsRegistry = {
${r.map(r=>({id:JSON.stringify(r.id),pkg:JSON.stringify(joinURL(`likec4:icons`,r.id))})).map(({id:r,pkg:i})=>`${r}: lazy(() => import(${i}).then(m => ({default: m.IconRenderer})))`).join(`,
`)}
}      


export function getProjectIcons(projectId) {
  let fn = ProjectIconsRegistry[projectId]
  if (!fn) {
    const projects = Object.keys(ProjectIconsRegistry)
    console.error('Unknown projectId: ' + projectId + ' (available: ' + projects + ')')
    if (projects.length === 0) {
      throw new Error('No projects found, invalid state')
    }
    projectId = projects[0]
    console.warn('Falling back to project: ' + projectId)
    fn = ProjectIconsRegistry[projectId]
  }
  return (props) => jsx(Suspense, { children: jsx(fn, props) })
}

if (import.meta.hot) {
  import.meta.hot.accept(md => {
    if (!import.meta.hot.data.$update) {
      import.meta.hot.data.$update = ProjectIconsRegistry
    }
    const update = md.ProjectIconsRegistry
    if (update) {
      Object.assign(import.meta.hot.data.$update, update)
    } else {
      import.meta.hot.invalidate()
    }
  })
}
`}};function code$5(r){let i=new CompositeGeneratorNode;return i.appendTemplate`
    /******************************************************************************
     * This file was generated
     * DO NOT EDIT MANUALLY!
     ******************************************************************************/
    /* eslint-disable */

    export function mmdSource(viewId) {
      switch (viewId) {
  `.appendNewLine().indent({indentation:4,indentedChildren(i){i.append(joinToNode([...r.views()],r=>expandToNode`
              case ${JSON.stringify(r.id)}: {
                return ${JSON.stringify(generateMermaid(r))}
              }
            `,{appendNewLineIfNotEmpty:!0})).appendTemplate`
          default: {
            throw new Error('Unknown viewId: ' + viewId)
          }
        `}}).append(NL,`  }`,NL).appendTemplate`
    }

    `.append(NL,NL),toString(i)}const projectMmdSourcesModule={...generateMatches(`mmd`),async load({likec4:r,project:i,logger:a}){return a.info(u.dim(`generating virtual:likec4/mmd/${i.id}`)),code$5(await r.computedModel(i.id))}},mmdModule=generateCombinedProjects(`mmd`,`loadMmdSources`),projectModelCode=r=>`
import { createHooksForModel, atom } from 'likec4/vite-plugin/internal'

export const $likec4data = atom(${lib.stringify(r.$data)})

export const {
  updateModel,
  $likec4model,
  useLikeC4Model,
  useLikeC4Views,
  useLikeC4View
} = createHooksForModel($likec4data)

if (import.meta.hot) {
  import.meta.hot.accept(md => {
    if (!import.meta.hot.data.$update) {
      import.meta.hot.data.$update = updateModel
    }
    const update = md.$likec4data?.get()
    if (update) {
      import.meta.hot.data.$update(update)
    } else {
      import.meta.hot.invalidate()
    }
  })
}
`,projectModelModule={...generateMatches(`model`),async load({likec4:r,project:i,logger:a}){return a.info(u.dim(`generating likec4:model/${i.id}`)),projectModelCode(await r.layoutedModel(i.id))}},modelModule=generateCombinedProjects(`model`,`loadModel`),code$4=r=>`
import { atom, useStore } from 'likec4/vite-plugin/internal'

export const isSingleProject = ${r.length===1};
export const projects = ${lib.stringify(r,null,2)};

export const $projects = atom([...projects])

export function useLikeC4Projects() {
  return useStore($projects)
}

if (import.meta.hot) {
  import.meta.hot.accept(md => {
    if (!import.meta.hot.data.$projects) {
      import.meta.hot.data.$projects = $projects
    }
    if (!import.meta.hot.data.projects) {
      import.meta.hot.data.projects = projects
    }      
    const update = md.projects
    if (update) {
      import.meta.hot.data.projects.length = 0
      import.meta.hot.data.projects.push(...update)
      import.meta.hot.data.$projects.set(update)
    } else {
      import.meta.hot.invalidate()
    }
  })
}
`,projectsModule={id:`likec4:projects`,virtualId:`likec4:plugin/projects.js`,async load({logger:r,projects:i}){return r.info(u.dim(`generating likec4:projects`)),code$4(t$3(i,t$4([`id`,`title`])))}},code$3=r=>`
import { atom, useStore } from 'likec4/vite-plugin/internal'

export const $viewdata = atom(${lib.stringify(r)})

export function useLikeC4ProjectsOverview() {
  return useStore($viewdata)
}

if (import.meta.hot) {
  import.meta.hot.accept(md => {
    if (!import.meta.hot.data.$viewdata) {
      import.meta.hot.data.$viewdata = $viewdata
    }
    const update = md.$viewdata?.value
    if (update) {
      import.meta.hot.data.$viewdata.set(update)
    } else {
      import.meta.hot.invalidate()
    }
  })
}
`,noProjects=()=>`
export function useLikeC4ProjectsOverview() {
  throw new Error('No projects overview available for this workspace: single project mode is enabled')
}
`,projectsOverviewModule={id:`likec4:projects-overview`,virtualId:`likec4:plugin/projects-overview.js`,async load({logger:r,projects:i,likec4:a}){return i.length<2?noProjects():(r.info(u.dim(`generating likec4:projects-overview`)),code$3(await a.projectsOverview()))}};function code$2(r){let i=new CompositeGeneratorNode;return i.appendTemplate`
    /******************************************************************************
     * This file was generated
     * DO NOT EDIT MANUALLY!
     ******************************************************************************/
    /* eslint-disable */

    export function pumlSource(viewId) {
      switch (viewId) {
  `.appendNewLine().indent({indentation:4,indentedChildren(i){i.append(joinToNode([...r.views()],r=>expandToNode`
              case ${JSON.stringify(r.id)}: {
                return ${JSON.stringify(generatePuml(r))}
              }
            `,{appendNewLineIfNotEmpty:!0})).appendTemplate`
        default: {
          throw new Error('Unknown viewId: ' + viewId)
        }
      `}}).append(NL,`  }`,NL).appendTemplate`
    }
    `.append(NL,NL),toString(i)}const projectPumlModule={...generateMatches(`puml`),async load({likec4:r,project:i,logger:a}){return a.info(u.dim(`generating likec4:puml/${i.id}`)),code$2(await r.computedModel(i.id))}},pumlModule=generateCombinedProjects(`puml`,`loadPumlSources`),projectCode=r=>`
import { jsx as _jsx } from "react/jsx-runtime";
import { LikeC4ModelProvider as Provider, LikeC4View as GenericView, ReactLikeC4 as GenericReactLikeC4 } from 'likec4/react';
import { IconRenderer } from 'likec4:icons/${r}'

import { useLikeC4Model, useLikeC4Views, useLikeC4View } from 'likec4:model/${r}'

export function LikeC4ModelProvider({ children }) {
  const likeC4Model = useLikeC4Model()
  return (_jsx(Provider, { likec4model: likeC4Model, children: children }));
}
export function LikeC4View(props) {
  return (_jsx(LikeC4ModelProvider, { children: _jsx(GenericView, { renderIcon: IconRenderer, ...props }) }));
}
export function ReactLikeC4(props) {
  return (_jsx(LikeC4ModelProvider, { children: _jsx(GenericReactLikeC4, { renderIcon: IconRenderer, ...props }) }));
}

export {
  useLikeC4Model,
  useLikeC4View,
  useLikeC4Views
}
`,projectReactModule={...generateMatches(`react`),async load({project:r,logger:i}){return i.info(u.dim(`generating likec4:react/${r.id}`)),projectCode(r.id)}},singleProjectReactModule={id:`likec4:react`,virtualId:`likec4:plugin/react.js`,async load({logger:r,projects:i}){let a=n$2(i);return r.info(u.dim(`generating likec4:react for`)+` `+a.id),projectCode(a.id)}},rpcModule={id:`likec4:rpc`,virtualId:`likec4:plugin/rpc.js`,async load({logger:r}){return r.info(u.dim(`generating likec4:rpc`)),` 
import { createRpc } from 'likec4/vite-plugin/internal'

export const isRpcAvailable = !!import.meta.hot

let rpc 
if (import.meta.hot) {
  rpc = createRpc({
    send: (event, data) => {
      import.meta.hot.send(event, data)
    },
    on: (event, fn) => {
      import.meta.hot.on(event, fn)
    }
  })

  import.meta.hot.accept()
}

export const likec4rpc = rpc ?? {
  updateView: () => {
    throw new Error('likec4rpc.updateView is not available in production')
  },
  calcAdhocView: () => {
    throw new Error('likec4rpc.calcAdhocView is not available in production')
  },
}
`}},code=r=>`
export { IconRenderer } from 'likec4:icons/${r}'
export {
  $likec4data,
  $likec4model,
  useLikeC4Model,
  useLikeC4Views,
  useLikeC4View
} from 'likec4:model/${r}'
export const projectId = ${JSON.stringify(r)}
`,singleProjectModule={id:`likec4:single-project`,virtualId:`likec4:plugin/single-project.js`,async load({logger:r,projects:i}){let a=n$2(i);return r.info(u.dim(`generating likec4:single-project for`)+` `+a.id),code(a.id)}},hmrProjectVirtuals=[projectModelModule,projectIconsModule,projectD2Module,projectDotSourcesModule,projectMmdSourcesModule,projectPumlModule],projectVirtuals=[...hmrProjectVirtuals,projectReactModule],virtuals=[projectsModule,modelModule,projectsOverviewModule,singleProjectModule,singleProjectReactModule,d2Module,dotModule,mmdModule,pumlModule,iconsModule,rpcModule];function LikeC4VitePlugin({environments:r,...i}){let o,s,c=i.watch??!1;return{name:`vite-plugin-likec4`,applyToEnvironment(i){return r?r.includes(i.name):!0},async configResolved(r){if(i.languageServices)o=i.languageServices;else{let a=c=i.watch??r.mode===`development`;o=(await fromWorkspace(i.workspace??r.root,{graphviz:i.graphviz??`wasm`,watch:a})).languageServices}s=o.workspaceUri.fsPath},resolveId(r){for(let i of projectVirtuals){let a=i.matches(r);if(a)return i.virtualId(a)}for(let i of virtuals)if(i.id===r)return i.virtualId;return null},async load(r){for(let i of projectVirtuals){let a=i.matches(r);if(a){let r=o.project(a);return await i.load({logger,likec4:o,project:r,assetsDir:s})}}for(let i of virtuals)if(i.virtualId===r){let r=o.projects();return invariant(isNonEmptyArray(r)),await i.load({logger,likec4:o,projects:r,assetsDir:s})}return null},configureServer(r){let i=r.hot;enablePluginRPC.call(this,{logger,likec4:o,server:r});let s=()=>t$3(o.projects(),r=>({id:r.id,title:r.title,folder:r.folder.toString()})),c=s(),l=async i=>{let a=r.moduleGraph.getModuleById(i);if(a&&a.importers.size>0)try{await r.reloadModule(a)}catch(r){logger.error(loggable(r))}};o.builder.onModelParsed(async()=>{let[r]=o.getErrors();if(r){i.send({type:`error`,err:{name:`LikeC4ValidationError`,...splitErrorMessage(r.message),plugin:`vite-plugin-likec4`,loc:{file:r.sourceFsPath,line:r.line,column:r.range.start.character+1}}});return}let d=s();if(!t$5(d,c)){c=d,await l(projectsModule.virtualId),await l(iconsModule.virtualId),await l(modelModule.virtualId),c.length>1&&await l(projectsOverviewModule.virtualId);return}for(let r of d)for(let i of hmrProjectVirtuals)await l(i.virtualId(r.id))})},async buildEnd(){c&&await o.dispose()}}}export{fromWorkspace as a,generateD2 as i,generatePuml as n,generateMermaid as r,LikeC4VitePlugin as t};
