// likec4/__app__/src/routes/index.js
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { cY as useNavigate, bS as useCallbackRef, cZ as z, c_ as createRootRouteWithContext, a0 as t, c$ as stripSearchParams, d0 as Outlet, k as useMantineColorScheme, d1 as createFileRoute, d2 as redirect, d3 as useDocumentTitle, d4 as createRouter, d5 as createHashHistory, d6 as createBrowserHistory, d7 as RouterProvider } from "../vendors.js";
import { useLikeC4Projects, projects } from "likec4:projects";
import { R as Route$3, a as Route$4, b as Route$5, c as Route$9, d as Route$a, e as Route$b, f as Route$c, g as Route$d, h as Route$h, i as Route$i, j as Route$j, k as Route$k, F as Fallback, N as NotFound } from "./single.js";
import { pageTitle, useHashHistory, basepath } from "../const.js";
import { L as LikeC4ProjectsProvider, B as Box } from "../likec4.js";
import { useEffect, lazy } from "react";
import { R as Route$6, a as Route$7, b as Route$8, c as Route$e, d as Route$f, e as Route$g, f as Route$l, g as Route$m, h as Route$n, i as Route$o, j as Route$p } from "./projects.js";
function LikeC4ProjectsContext({ children }) {
  const projects2 = useLikeC4Projects(), navigate = useNavigate(), onProjectChange = useCallbackRef((projectId) => {
    navigate({ to: "/project/$projectId/", params: { projectId } }).catch(console.error);
  });
  return /* @__PURE__ */ jsx(
    LikeC4ProjectsProvider,
    {
      projects: projects2,
      onProjectChange,
      children
    }
  );
}
const searchParamsSchema = z.object({
  theme: z.literal(["light", "dark", "auto"]).default("auto").catch("auto"),
  dynamic: z.enum(["diagram", "sequence"]).default("diagram").catch("diagram"),
  padding: z.number().min(0).default(20).catch(20),
  relationships: z.string().nonempty().optional().catch(void 0).transform((v) => v)
}), Route$2 = createRootRouteWithContext()({
  validateSearch: searchParamsSchema,
  search: {
    middlewares: [
      stripSearchParams({
        padding: 20,
        theme: "auto",
        dynamic: "diagram",
        relationships: void 0
      })
    ]
  },
  beforeLoad: () => {
    const _projects = projects.length > 0 ? t(projects, (p) => p.id) : ["default"];
    return {
      projects: _projects,
      projectId: _projects[0]
    };
  },
  component: RootComponent
});
function RootComponent() {
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Outlet, {}),
    /* @__PURE__ */ jsx(ThemeSync, {})
  ] });
}
const ThemeSync = () => {
  const { theme } = Route$2.useSearch(), mantineColorScheme = useMantineColorScheme();
  return useEffect(() => {
    theme && mantineColorScheme.setColorScheme(theme);
  }, [theme]), null;
}, ProjectsOverviewPage = lazy(async () => {
  const { ProjectsOverviewPage: ProjectsOverviewPage2 } = await import("../ProjectsOverview.js");
  return {
    default: ProjectsOverviewPage2
  };
}), Route$1 = createFileRoute("/projects")({
  beforeLoad: async ({ context }) => {
    if (context.projects.length < 2)
      throw redirect({
        to: "/single-index/"
      });
  },
  component: RouteComponent,
  wrapInSuspense: !0
});
function RouteComponent() {
  return useDocumentTitle(`Projects - ${pageTitle}`), /* @__PURE__ */ jsx(Box, { w: "100%", h: "100%", overflow: "hidden", children: /* @__PURE__ */ jsx(ProjectsOverviewPage, {}) });
}
const Route = createFileRoute("/")({
  beforeLoad: ({ context }) => {
    throw context.projects.length > 1 ? redirect({
      to: "/projects/",
      mask: {
        to: "/",
        unmaskOnReload: !0
      }
    }) : redirect({
      to: "/single-index/",
      mask: {
        to: "/",
        unmaskOnReload: !0
      }
    });
  }
}), ProjectsRoute = Route$1.update({
  id: "/projects",
  path: "/projects",
  getParentRoute: () => Route$2
}), SingleRouteRoute = Route$3.update({
  id: "/_single",
  getParentRoute: () => Route$2
}), IndexRoute = Route.update({
  id: "/",
  path: "/",
  getParentRoute: () => Route$2
}), SingleSingleIndexRoute = Route$4.update({
  id: "/single-index",
  path: "/single-index",
  getParentRoute: () => SingleRouteRoute
}), SingleAdhocRoute = Route$5.update({
  id: "/adhoc",
  path: "/adhoc",
  getParentRoute: () => SingleRouteRoute
}), ProjectProjectIdRouteRoute = Route$6.update({
  id: "/project/$projectId",
  path: "/project/$projectId",
  getParentRoute: () => Route$2
}), ProjectProjectIdIndexRoute = Route$7.update({
  id: "/",
  path: "/",
  getParentRoute: () => ProjectProjectIdRouteRoute
}), ProjectProjectIdAdhocRoute = Route$8.update({
  id: "/adhoc",
  path: "/adhoc",
  getParentRoute: () => ProjectProjectIdRouteRoute
}), SingleWebcomponentSplatRoute = Route$9.update({
  id: "/webcomponent/$",
  path: "/webcomponent/$",
  getParentRoute: () => SingleRouteRoute
}), SingleViewViewIdRoute = Route$a.update({
  id: "/view/$viewId",
  path: "/view/$viewId",
  getParentRoute: () => SingleRouteRoute
}), SingleExportViewIdRoute = Route$b.update({
  id: "/export/$viewId",
  path: "/export/$viewId",
  getParentRoute: () => SingleRouteRoute
}), SingleEmbedViewIdRoute = Route$c.update({
  id: "/embed/$viewId",
  path: "/embed/$viewId",
  getParentRoute: () => SingleRouteRoute
}), SingleViewViewIdIndexRoute = Route$d.update({
  id: "/",
  path: "/",
  getParentRoute: () => SingleViewViewIdRoute
}), ProjectProjectIdViewViewIdRoute = Route$e.update(
  {
    id: "/view/$viewId",
    path: "/view/$viewId",
    getParentRoute: () => ProjectProjectIdRouteRoute
  }
), ProjectProjectIdExportViewIdRoute = Route$f.update({
  id: "/export/$viewId",
  path: "/export/$viewId",
  getParentRoute: () => ProjectProjectIdRouteRoute
}), ProjectProjectIdEmbedViewIdRoute = Route$g.update({
  id: "/embed/$viewId",
  path: "/embed/$viewId",
  getParentRoute: () => ProjectProjectIdRouteRoute
}), SingleViewViewIdPumlRoute = Route$h.update({
  id: "/puml",
  path: "/puml",
  getParentRoute: () => SingleViewViewIdRoute
}), SingleViewViewIdMmdRoute = Route$i.update({
  id: "/mmd",
  path: "/mmd",
  getParentRoute: () => SingleViewViewIdRoute
}), SingleViewViewIdDotRoute = Route$j.update({
  id: "/dot",
  path: "/dot",
  getParentRoute: () => SingleViewViewIdRoute
}), SingleViewViewIdD2Route = Route$k.update({
  id: "/d2",
  path: "/d2",
  getParentRoute: () => SingleViewViewIdRoute
}), ProjectProjectIdViewViewIdIndexRoute = Route$l.update({
  id: "/",
  path: "/",
  getParentRoute: () => ProjectProjectIdViewViewIdRoute
}), ProjectProjectIdViewViewIdPumlRoute = Route$m.update({
  id: "/puml",
  path: "/puml",
  getParentRoute: () => ProjectProjectIdViewViewIdRoute
}), ProjectProjectIdViewViewIdMmdRoute = Route$n.update({
  id: "/mmd",
  path: "/mmd",
  getParentRoute: () => ProjectProjectIdViewViewIdRoute
}), ProjectProjectIdViewViewIdDotRoute = Route$o.update({
  id: "/dot",
  path: "/dot",
  getParentRoute: () => ProjectProjectIdViewViewIdRoute
}), ProjectProjectIdViewViewIdD2Route = Route$p.update({
  id: "/d2",
  path: "/d2",
  getParentRoute: () => ProjectProjectIdViewViewIdRoute
}), SingleViewViewIdRouteChildren = {
  SingleViewViewIdD2Route,
  SingleViewViewIdDotRoute,
  SingleViewViewIdMmdRoute,
  SingleViewViewIdPumlRoute,
  SingleViewViewIdIndexRoute
}, SingleViewViewIdRouteWithChildren = SingleViewViewIdRoute._addFileChildren(SingleViewViewIdRouteChildren), SingleRouteRouteChildren = {
  SingleAdhocRoute,
  SingleSingleIndexRoute,
  SingleEmbedViewIdRoute,
  SingleExportViewIdRoute,
  SingleViewViewIdRoute: SingleViewViewIdRouteWithChildren,
  SingleWebcomponentSplatRoute
}, SingleRouteRouteWithChildren = SingleRouteRoute._addFileChildren(
  SingleRouteRouteChildren
), ProjectProjectIdViewViewIdRouteChildren = {
  ProjectProjectIdViewViewIdD2Route,
  ProjectProjectIdViewViewIdDotRoute,
  ProjectProjectIdViewViewIdMmdRoute,
  ProjectProjectIdViewViewIdPumlRoute,
  ProjectProjectIdViewViewIdIndexRoute
}, ProjectProjectIdViewViewIdRouteWithChildren = ProjectProjectIdViewViewIdRoute._addFileChildren(
  ProjectProjectIdViewViewIdRouteChildren
), ProjectProjectIdRouteRouteChildren = {
  ProjectProjectIdAdhocRoute,
  ProjectProjectIdIndexRoute,
  ProjectProjectIdEmbedViewIdRoute,
  ProjectProjectIdExportViewIdRoute,
  ProjectProjectIdViewViewIdRoute: ProjectProjectIdViewViewIdRouteWithChildren
}, ProjectProjectIdRouteRouteWithChildren = ProjectProjectIdRouteRoute._addFileChildren(
  ProjectProjectIdRouteRouteChildren
), rootRouteChildren = {
  IndexRoute,
  SingleRouteRoute: SingleRouteRouteWithChildren,
  ProjectsRoute,
  ProjectProjectIdRouteRoute: ProjectProjectIdRouteRouteWithChildren
}, routeTree = Route$2._addFileChildren(rootRouteChildren)._addFileTypes(), router = createRouter({
  routeTree,
  context: {
    projectId: projects[0].id,
    projects: t(projects, (p) => p.id)
  },
  InnerWrap: LikeC4ProjectsContext,
  basepath,
  trailingSlash: "always",
  defaultViewTransition: !1,
  history: useHashHistory ? createHashHistory() : createBrowserHistory(),
  defaultStaleTime: 1 / 0,
  defaultNotFoundComponent: () => /* @__PURE__ */ jsx(NotFound, {}),
  defaultErrorComponent: ({ error, reset }) => /* @__PURE__ */ jsx(Fallback, { error, resetErrorBoundary: reset })
});
function Routes() {
  return /* @__PURE__ */ jsx(RouterProvider, { router });
}
export {
  Routes as R
};
