// likec4/__app__/src/routes/projects.js
import { jsx, jsxs, Fragment } from "react/jsx-runtime";
import { d1 as createFileRoute, db as Container, bp as Stack, bZ as Title, W as Button, df as Link, dD as notFound, d2 as redirect, Q as m, d0 as Outlet, dE as Navigate, c$ as stripSearchParams, cZ as z } from "../vendors.js";
import { loadModel } from "likec4:model";
import { V as ViewOutlet, F as Fallback, L as LikeC4IconRendererContext, l as LikeC4ModelContext, A as AdHocViewEditor, H as Header, E as ExportPage, m as EmbedPage, n as ViewEditor, o as ViewReact, p as ViewAsPuml, q as ViewAsMmd, r as ViewAsDot, s as ViewAsD2 } from "./single.js";
import { isRpcAvailable } from "likec4:rpc";
const Route$a = createFileRoute("/project/$projectId")({
  staleTime: 1 / 0,
  beforeLoad: ({ params }) => ({
    projectId: params.projectId
  }),
  loaderDeps() {
    return [];
  },
  loader: async ({ context }) => {
    const projectId = context.projectId, likec4model = await loadModel(projectId), data = likec4model.$likec4data.value;
    if (!data)
      throw notFound();
    if (data.projectId !== projectId)
      throw redirect({
        to: "/project/$projectId/",
        search: !0,
        params: {
          projectId: data.projectId
        }
      });
    return {
      $likec4model: likec4model.$likec4model,
      projectId
    };
  },
  remountDeps({ params }) {
    return [params.projectId];
  },
  component: RouteComponent,
  notFoundComponent: () => /* @__PURE__ */ jsx(ViewOutlet, { children: /* @__PURE__ */ jsx(Container, { py: "xl", children: /* @__PURE__ */ jsxs(Stack, { align: "flex-start", children: [
    /* @__PURE__ */ jsx(Title, { children: "Project not found" }),
    /* @__PURE__ */ jsx(Button, { component: Link, to: "/", search: !0, size: "md", children: "Open overview" })
  ] }) }) })
});
function RouteComponent() {
  const { $likec4model, projectId } = Route$a.useLoaderData();
  return /* @__PURE__ */ jsx(ViewOutlet, { children: /* @__PURE__ */ jsx(m, { FallbackComponent: Fallback, children: /* @__PURE__ */ jsx(LikeC4IconRendererContext, { projectId, children: /* @__PURE__ */ jsx(LikeC4ModelContext, { likec4model: $likec4model, children: /* @__PURE__ */ jsx(m, { FallbackComponent: Fallback, children: /* @__PURE__ */ jsx(Outlet, {}) }) }) }) }) });
}
const Route$9 = createFileRoute("/project/$projectId/")({
  component: () => {
    const { projectId } = Route$9.useParams();
    return /* @__PURE__ */ jsx(
      Navigate,
      {
        to: "/project/$projectId/view/$viewId/",
        replace: !0,
        params: {
          projectId,
          viewId: "index"
        }
      }
    );
  }
}), Route$8 = createFileRoute("/project/$projectId/adhoc")({
  beforeLoad: ({ params }) => {
    if (!isRpcAvailable)
      throw redirect({
        to: "/project/$projectId/",
        params: {
          projectId: params.projectId
        }
      });
  },
  component: Page$4
});
function Page$4() {
  const { projectId } = Route$8.useRouteContext();
  return /* @__PURE__ */ jsx(AdHocViewEditor, { projectId });
}
const Route$7 = createFileRoute("/project/$projectId/view/$viewId")({
  component: ViewLayout,
  errorComponent: ({ error, reset }) => /* @__PURE__ */ jsx(Fallback, { error, resetErrorBoundary: reset })
});
function ViewLayout() {
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Outlet, {}),
    /* @__PURE__ */ jsx(Header, {})
  ] });
}
const Route$6 = createFileRoute("/project/$projectId/export/$viewId")({
  validateSearch: z.object({
    download: z.boolean().optional().catch(!1)
  }),
  search: {
    middlewares: [
      stripSearchParams({
        download: !1
      })
    ]
  },
  component: ExportPage
}), Route$5 = createFileRoute("/project/$projectId/embed/$viewId")({
  component: EmbedPage
}), Route$4 = createFileRoute("/project/$projectId/view/$viewId/")({
  component: isRpcAvailable ? ViewEditor : ViewReact
}), Route$3 = createFileRoute("/project/$projectId/view/$viewId/puml")({
  component: Page$3,
  staleTime: 1 / 0,
  loader: async ({ params, context }) => {
    const projectId = context.projectId, { viewId } = params, { loadPumlSources } = await import("likec4:puml");
    try {
      const { pumlSource } = await loadPumlSources(projectId);
      return {
        source: pumlSource(viewId)
      };
    } catch (error) {
      throw console.error(error), notFound();
    }
  }
});
function Page$3() {
  const { source } = Route$3.useLoaderData();
  return /* @__PURE__ */ jsx(ViewAsPuml, { pumlSource: source });
}
const Route$2 = createFileRoute("/project/$projectId/view/$viewId/mmd")({
  component: Page$2,
  staleTime: 1 / 0,
  loader: async ({ params, context }) => {
    const projectId = context.projectId, { viewId } = params, { loadMmdSources } = await import("likec4:mmd");
    try {
      const { mmdSource } = await loadMmdSources(projectId);
      return {
        source: mmdSource(viewId)
      };
    } catch (error) {
      throw console.error(error), notFound();
    }
  }
});
function Page$2() {
  const { viewId } = Route$2.useParams(), { source } = Route$2.useLoaderData();
  return /* @__PURE__ */ jsx(ViewAsMmd, { viewId, mmdSource: source });
}
const Route$1 = createFileRoute("/project/$projectId/view/$viewId/dot")({
  component: Page$1,
  staleTime: 1 / 0,
  loader: async ({ params, context }) => {
    const projectId = context.projectId, { viewId } = params, { loadDotSources } = await import("likec4:dot");
    try {
      const { dotSource, svgSource } = await loadDotSources(projectId), dot = dotSource(viewId), dotSvg = svgSource(viewId);
      return {
        dot,
        dotSvg
      };
    } catch (error) {
      throw console.error(error), notFound();
    }
  }
});
function Page$1() {
  const { dot, dotSvg } = Route$1.useLoaderData();
  return /* @__PURE__ */ jsx(ViewAsDot, { dot, dotSvg });
}
const Route = createFileRoute("/project/$projectId/view/$viewId/d2")({
  loader: async ({ context, params }) => {
    const projectId = context.projectId, { viewId } = params, { loadD2Sources } = await import("likec4:d2");
    try {
      const { d2Source } = await loadD2Sources(projectId);
      return {
        source: d2Source(viewId)
      };
    } catch (error) {
      throw console.error(error), notFound();
    }
  },
  component: Page
});
function Page() {
  const { source } = Route.useLoaderData();
  return /* @__PURE__ */ jsx(ViewAsD2, { d2Source: source });
}
export {
  Route$a as R,
  Route$9 as a,
  Route$8 as b,
  Route$7 as c,
  Route$6 as d,
  Route$5 as e,
  Route$4 as f,
  Route$3 as g,
  Route$2 as h,
  Route$1 as i,
  Route as j
};
